File: data/two-pointers/sort-colors.cpp
#include <iostream>
#include <vector>
using namespace std;

void sortColors(vector<int>& nums) {
    int low = 0, mid = 0, high = nums.size() - 1;  // Initialize three pointers
    
    while (mid <= high) {  // Traverse the array
        if (nums[mid] == 0) {
            // Swap 0 to the 'low' position
            swap(nums[low], nums[mid]);
            low++;
            mid++;
        } else if (nums[mid] == 1) {
            // If it's 1, just move forward
            mid++;
        } else {
            // Swap 2 to the 'high' position
            swap(nums[mid], nums[high]);
            high--;
        }
    }
}

// Example Usage
int main() {
    vector<int> nums = {2, 0, 2, 1, 1, 0};
    sortColors(nums);
    
    // Output sorted array
    for (int num : nums) {
        cout << num << " ";  // Output: 0 0 1 1 2 2
    }
    cout << endl;
    
    return 0;
}

File: data/two-pointers/sort-colors.py
def sortColors(nums):
    """
    Sorts the array in-place such that all 0's come first, followed by all 1's, and then all 2's.
    
    :param nums: List[int] - List containing the integers 0, 1, and 2
    """
    low, mid, high = 0, 0, len(nums) - 1  # Initialize three pointers
    
    while mid <= high:  # Traverse through the array
        if nums[mid] == 0:
            # Swap 0 to the 'low' position
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif nums[mid] == 1:
            # If it's 1, just move forward
            mid += 1
        else:
            # Swap 2 to the 'high' position
            nums[high], nums[mid] = nums[mid], nums[high]
            high -= 1
    
# Example Usage
nums = [2, 0, 2, 1, 1, 0]
sortColors(nums)
print(nums)  # Output: [0, 0, 1, 1, 2, 2]
